[KOGITO-7288] [KSW-Guides] Event correlation with workflows (#85)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/eventing/event-correlation-with-workflows.adoc

= Event correlation with workflows
=====================================
:compat-mode!:
// Metadata:
:description: Event Correlation in Serverless Workflow
:keywords: kogito, workflow, serverless, correlation, association, domain, events

[[ref-example-jq-expression-switch-conditions]]

Event correlation plays a big role in large event-driven applications, it allows to match one or more events with a particular workflow instance. Correlation  rules can be used when defining consumed events as a way to match a workflow instance, it is a more practical alternative to the internal identifier `processInstanceId`, since any external domain identifier can be used as a correlation.

> A correlation definition is consisted of one or more attributes associated with an event and the respective workflow to which it belongs.  As described in the link:{spec_doc_url}#correlation-definition[Serverless Workflow Specification], the correlation property defines the possible correlations for a given event, each element should contain a `contextAttributeName` set for the value that should match an attribute from an event. Optionally, a `contextAttributeValue` property defines a constant value that should be used to match an event instead of the value present in the consumed event, during the workflow instance creation.

Reviewer: A correlation definition consists of?

Author: :+1:

[[ref-new-user-account-state]]
.`New User Account Request` state definition
[source,json]
----
{
  "name": "New User Account Request",
  "type": "event",
  "onEvents": [
    {
      "eventRefs": [
        "newAccountEvent"
      ]
    }
  ],
  "transition": "Validate User Email"
}
----

When a new event of type <<ref-new-user-account-event,`newAccountEventType`>> is consumed the workflow is created and then any other events consumed by this workflow should contain the same correlation attribute `userid` with the same value `12345` set, which is used to evaluate and match the workflow instance to continue the execution.

[[ref-new-user-account-event]]
.New Incoming Start Event `newAccountEvent`
[source,json]
----
{
  "specversion": "0.3",
  "id": "1d174d25-46ac-4785-bc76-457c2d37d2fe",
  "source": "",
  "type": "newAccountEventType",
  "time": "2022-07-25T16:30:35.461988261-03:00",
  "userid": "12345",
  "data": {
    "email": "test@test.com",
    "userId": "12345"
  }
}
----

[NOTE]
====
It is not supported to correlate multiple events together, events are evaluated with correlations independently.
====

=== Producing and consuming events with correlation

> Following the link:{kogito_sw_examples_url}/serverless-workflow-correlation-quarkus/src/main/resources/correlation.sw.json[workflow definition], it is possible to define other events that are supposed to be published and consumed by this workflow, given example uses Callback States, such as the  <<ref-validade-user-email-state,`Validate User Email`>>, indicating once the workflow execution achieves this state, it publishes an event of type <<ref-correlation-event-definition,`validateAccountEmailEvent`>> and wait to receive an event of type <<ref-correlation-event-definition,`validatedAccountEmailEvent`>>, more information on xref:eventing/working-with-callbacks.adoc[Working with callbacks].

Reviewer: I guess Heena will say that, but I think you can use more "." rather than "," in this parragraph ;)

Author: :+1:

= Event correlation with workflows
=====================================
:compat-mode!:
// Metadata:
:description: Event Correlation in Serverless Workflow
:keywords: kogito, workflow, serverless, correlation, association, domain, events

[[ref-example-jq-expression-switch-conditions]]

> Event correlation plays a big role in large event-driven applications, it allows to match one or more events with a particular workflow instance. Correlation  rules can be used when defining consumed events as a way to match a workflow instance, it is a more practical alternative to the internal identifier `processInstanceId`, since any external domain identifier can be used as a correlation.

Reviewer 2: ```suggestion
Event correlation plays a significant role in large event-driven applications. It allows matching one or more events with a particular workflow instance. Correlation rules with domain information can be used when defining consumed events to match a workflow instance. It is a more practical alternative to the internal identifier `processInstanceId` since you can use any external domain identifier as a correlation.
```

Author: +1

> A correlation definition consists of one or more attributes associated with an event and the respective workflow to which it belongs.  As described in the link:{spec_doc_url}#correlation-definition[Serverless Workflow Specification], the correlation property defines the possible correlations for a given event, each element should contain a `contextAttributeName` set for the value that should match an attribute from an event.

Reviewer 2: ```suggestion
A correlation definition consists of one or more attributes associated with an event and the respective workflow to which it belongs.  As described in the link:{spec_doc_url}#correlation-definition[Serverless Workflow Specification], the correlation property defines the possible correlations for a given event. Each element should contain a `contextAttributeName` set for the value that should match an attribute from an event.
```

Author: +1

Author: +1

> Optionally, you can set a `contextAttributeValue` property that should match the value for the respective attribute defined in `contextAttributeName` for the consumed events in a workflow.

Reviewer 2: ```suggestion
You can optionally set a `contextAttributeValue` property that should match the value for the respective attribute defined in `contextAttributeName` for the consumed events in a workflow.
```

Can you add a workflow definition example to illustrate this saying (don't need to add the entire workflow, only the portion that makes sense)? Assuming that the user won't click on the SW link.

Author: +1

Author: +1

[NOTE]
====
The incoming events consumed by the workflow engine should contain the correlation attributes, set in the definition, as extension context attributes, in compliance with the link:{cloud_events_url}[Cloud Event] format, thus they are not part of the event payload.
====

> When starting a new workflow instance, the correlation attributes set in the event definition, are extracted from the event and associated with the workflow instance which is being created. Thus, the start event does not trigger a correlation evaluation, instead, it is where correlation attributes and values are set, which are used to evaluate against other incoming events that might trigger this instance. Once an event is consumed and the correlation rules are evaluated, if any workflow instance matches them, the engine continues the execution of that matched instance.

Reviewer 2: I believe we can shorten this paragraph to make it clearer. But @hmanwani-rh would do a better job than me. 😂

The following flowchart diagram shows how the correlation is processed by the workflow engine.

.Event correlation evaluation process

> image::eventing/event-correlation-evaluation-flow.png[]

Reviewer 2: The "Is Start Event?" conditional is missing the labels for "Yes" and "No". Great illustration!

Author: +1

== Setting event correlation in serverless workflow

Correlation configuration can be found in the link:{kogito_sw_examples_url}/serverless-workflow-correlation-quarkus[serverless-workflow-correlation-quarkus] example, where it shows the setup and usage of event correlation in a workflow.

.Event correlation evaluation process

> image::eventing/correlation.sw.json.png[Event correlation evaluation process,500]

Reviewer 2: I know that's a lot to ask, but only if you have the time. Can you use the new Stunner viewer instead?

Author: +1

      "source": "workflow",
      "type": "validateAccountEmail"
    },
    {
      "name": "validatedAccountEmailEvent",
      "source": "workflow",
      "type": "validatedAccountEmail",
      "correlation": [
        {
          "contextAttributeName": "userid"
        }
      ]
    },
    {
      "name": "activateAccountEvent",
      "source": "workflow",
      "type": "activateAccount"
    },
    {
      "name": "activatedAccountEvent",
      "source": "workflow",
      "type": "activatedAccount",
      "correlation": [
        {
          "contextAttributeName": "userid"
        }
      ]
    }
  ]
----

A workflow can be created by consuming events as defined by the Event state <<ref-new-user-account-state,`New User Account Request`>> which has a reference to the event <<ref-correlation-event-definition,`newAccountEvent`>> containing a correlation definition for attribute `userid`.

[[ref-new-user-account-state]]
.`New User Account Request` state definition
[source,json]
----
{
  "name": "New User Account Request",
  "type": "event",
  "onEvents": [
    {
      "eventRefs": [
        "newAccountEvent"
      ]
    }
  ],
  "transition": "Validate User Email"
}
----

> When a new event of type <<ref-new-user-account-event,`newAccountEventType`>> is consumed the workflow is created and then any other events consumed by this workflow should contain the same correlation attribute `userid` with the same value `12345` set, which is used to evaluate and match the workflow instance to continue the execution.

Reviewer 2: ```suggestion
When the workflow consumes a new event of type <<ref-new-user-account-event,`newAccountEventType`>> the instance is created, and then any other events consumed by this workflow should contain the same correlation attribute `userid` with the same value `12345` set, which is used to evaluate and match the workflow instance to continue the execution.
```
I'd shorten this paragraph.

Author: +1

====

=== Producing and consuming events with correlation

Following the link:{kogito_sw_examples_url}/serverless-workflow-correlation-quarkus/src/main/resources/correlation.sw.json[workflow definition], it is possible to define other events that are supposed to be published and consumed by this workflow.

The given example uses Callback States, such as the  <<ref-validade-user-email-state,`Validate User Email`>>, indicating once the workflow execution achieves this state, it publishes an event of type <<ref-correlation-event-definition,`validateAccountEmailEvent`>> and wait to receive an event of type <<ref-correlation-event-definition,`validatedAccountEmailEvent`>>. For more information on xref:eventing/working-with-callbacks.adoc[Working with callbacks].

[[ref-validade-user-email-state]]
.Callback State Definition
[source,json]
----
{
  "name": "Validate User Email",
  "type": "callback",
  "action": {
    "name": "publish validate event",
    "eventRef": {
      "triggerEventRef": "validateAccountEmailEvent"
    }
  },
  "eventRef": "validatedAccountEmailEvent",
  "transition": "Activate User Account"
}
----

> [IMPORTANT]
> The produced events contain the same correlation attributes set once the workflow was created.

Reviewer 2: ```suggestion
[IMPORTANT]
===
The produced events contain the same correlation attributes set once the workflow was created.
===
```

Author: +1

`userid` with value `12345`, see <<ref-validaded-user-email-event,validatedAccountEmailEvent>>.

[[ref-validaded-user-email-event]]
.Consumed Callback State event `validatedAccountEmailEvent`
[source,json]
----
{
  "specversion": "1.0",
  "id": "953f07a7-aea8-4956-8775-85ab59366fe6",
  "source": "",
  "type": "validatedAccountEmail",
  "time": "2022-07-25T16:29:27.320408379-03:00",
  "userid": "12345",
  "data": null
}
----

== Correlation configuration

The workflow engine stores the correlation information in the same persistence mechanism configured in the application. If no persistence addon is configured, the correlation information is stored in memory, this means,  all correlation information is lost upon the application restart, so it should only be used for testing purposes. For more information about persistence configuration, see xref:persistence/persistence-with-postgresql.adoc[Persistence with PostgresSQL databases].

[NOTE]
====
Currently, the only persistency addon which supports correlation is the `kogito-addons-quarkus-persistence-jdbc` configured for PostgreSQL. In the next releases, other persistency addons will be supported.

> ====

Reviewer 2: Can you please include the "Found an issue" link and the "Additional resources" section?

Author: +1

A correlation definition consists of one or more attributes associated with an event and the respective workflow to which it belongs.  As described in the link:{spec_doc_url}#correlation-definition[Serverless Workflow Specification], the correlation property defines the possible correlations for a given event. Each element should contain a `contextAttributeName` set for the value that should match an attribute from an event.

You can optionally set a `contextAttributeValue` property that should match the value for the respective attribute defined in `contextAttributeName` for the consumed events in a workflow.

[NOTE]
====
The incoming events consumed by the workflow engine should contain the correlation attributes, set in the definition, as extension context attributes, in compliance with the link:{cloud_events_url}[Cloud Event] format, thus they are not part of the event payload.
====

> When starting a new workflow instance, the correlation attributes set in the event definition, are extracted from the event and associated with the workflow instance which is being created. Thus, the start event does not trigger a correlation evaluation, instead, it is where correlation attributes and values are set, which are used to evaluate against other incoming events that might trigger this instance.

Reviewer 3: I miss the information `who` starts this new workflow instance (as it is basically an engine itself). I understand it in a way that first an event is received by the service, then each workflow with its event and correlation definitions is tried to be matched against the received event and if there is a match, a new instance of the matching workflow is started and the correlation values are set. If there is an already running instance of the matched workflow, it continues the execution.

Author: The workflow is started once an event that is associated with the Event state set as a start in the workflow is received. So the who is the one publishing the event is similar to the one who might send a POST request to start a new workflow instance via REST...

Reviewer 3: Yes, my point was to maybe write it in the document so it is more obvious.

> When starting a new workflow instance, the correlation attributes set in the event definition, are extracted from the event and associated with the workflow instance which is being created. Thus, the start event does not trigger a correlation evaluation, instead, it is where correlation attributes and values are set, which are used to evaluate against other incoming events that might trigger this instance.

Reviewer 3: ```suggestion
When starting a new workflow instance, the correlation attributes set in the event definition are extracted from the event and associated with the workflow instance which is being created. Thus, the start event does not trigger a correlation evaluation, instead, it is where correlation attributes and values are set, which are used to evaluate against other incoming events that might trigger this instance.
```

> Once an event is consumed and the correlation rules are evaluated, if any workflow instance matches them, the engine continues the execution of that matched instance.

Reviewer 3: As I said previously, I would maybe add what happens before the actual start - that an event with the correlation attributes must be sent from somewhere and is received by the workflow service.

Author: Ok, I inserted one more sentence to clarify this part and a link to the Event State guide.
Thanks @MarianMacik

  "transition": "Activate User Account"
}
----

[IMPORTANT]
====
The produced events contain the same correlation attributes set once the workflow was created.
====

[[ref-validade-user-email-event]]
.Produced Callback State event `validateAccountEmailEvent`
[source,json]
----
{
  "id": "7640a0af-b7fb-4d94-9d9d-3aa1ace60e79",
  "source": "/process/correlation",
  "type": "validateAccountEmail",
  "time": "2022-07-25T16:22:53.735128049-03:00",
  "data": {
    "email": "test@test.com",
    "userId": "12345"
  },
  "specversion": "1.0",
  "kogitoprocinstanceid": "69019826-daef-4fb4-880b-c1658c4e49bc",
  "kogitoprocid": "correlation",
  "kogitoprocversion": "1.0",
  "kogitousertaskist": "1",
  "kogitoproctype": "SW",
  "userid": "12345"
}
----

All consumed events must contain the same correlation attributes since they are used to identify the workflow instance, in this example
`userid` with value `12345`, see <<ref-validaded-user-email-event,validatedAccountEmailEvent>>.

[[ref-validaded-user-email-event]]
.Consumed Callback State event `validatedAccountEmailEvent`
[source,json]
----
{
  "specversion": "1.0",
  "id": "953f07a7-aea8-4956-8775-85ab59366fe6",

>   "source": "",

Reviewer 3: What about the source? Doesn't it have to be set to `workflow` as is in the workflow definition? Or this is currently not matched at all?

Author: @MarianMacik 'source' property is not being considered to match when using correlation, at least on the current impl so in the payload the source could be any value... that's why I kept in this way in the example...

Reviewer 3: OK, thanks.

  "type": "validatedAccountEmail",
  "time": "2022-07-25T16:29:27.320408379-03:00",
  "userid": "12345",
  "data": null
}
----

== Correlation configuration

The workflow engine stores the correlation information in the same persistence mechanism configured in the application. If no persistence addon is configured, the correlation information is stored in memory, this means,  all correlation information is lost upon the application restart, so it should only be used for testing purposes. For more information about persistence configuration, see xref:persistence/persistence-with-postgresql.adoc[Persistence with PostgresSQL databases].

[NOTE]
====

> Currently, the only persistency addon which supports correlation is the `kogito-addons-quarkus-persistence-jdbc` configured for PostgreSQL. In the next releases, other persistency addons will be supported.

Reviewer 3: ```suggestion
Currently, the only persistence addon which supports correlation is the `kogito-addons-quarkus-persistence-jdbc` configured for PostgreSQL. In the next releases, other persistency addons will be supported.
```

Author: done

You can optionally set the `contextAttributeValue` property, which matches the value for the respective attribute defined in `contextAttributeName` property for the consumed events in a workflow.

[NOTE]
====
The incoming events consumed by the engine must contain the correlation attributes, set in the definition as extension context attributes. The correlation attributes are compliant with the link:{cloud_events_url}[CloudEvent] format, therefore, the attributes are not part of the event payload.
====

A new workflow instance must be created using an event, which must be declared in the workflow definition file, containing correlation attributes in the event definition section. For more information about events, see xref:eventing/handling-events-on-workflows.adoc[Event state in {context}]. Once the event is consumed, the engine extracts the correlation attributes and associates the attributes with the created workflow instance.

> A start event does not trigger a correlation evaluation, but acts as a moment in which correlation attributes and values are set. The correlation attributes and values are evaluated against other incoming events that might trigger the given instance. Therefore, when a non-start event is consumed and correlation attributes are evaluates, then the engine continues the execution of the matched instances (if any).

Author: evaluates -> evaluated ?

Reviewer 4: Done
